### DISCUSSION TITLE
workdir for custom_target

### DISCUSSION THREAD

**Discussion by georgmu** [mesonbuild/meson#10390] [Ideas]
Hi,

having a workdir parameter for custom_target() would be very handy.

I have a project which is a lot of C, but also some go, which currently uses cmake. Compiling the go stuff is done through custom targets.
Since `go build` only works in the source tree, but allows writing the output somewhere else, the custom target in cmake looks something like this:
```
add_custom_target(goprog
  COMMAND go build -mod=vendor -o ${CMAKE_CURRENT_BINARY_DIR}/goprog
  WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}
)
```

With meson, it is not possible to specify a workdir for such a scenario. The only alternative I see is to introduce a script which takes the souce and target directory as arguments and does the computation itself (using `readlink -f`).

I think such a change should be possible by using absolute paths to replace `@INPUT@`, `@OUTPUT@` and similar placeholders if `workdir` is set (instead of the `rel_to_builddir()`).

Any thoughts about this idea?

**Comment by mcb2003** [comment]
Another example of where this could be useful: I'm building a project that uses [Lua](https://lua.org), and documenting both the Lua and C code with [LDoc](https://github.com/lunarmodules/ldoc). I tried the following `custom_target` to build the documentation:
```meson
  docs = custom_target('docs',
    command: [ldoc, '-d', '@OUTDIR@/doc', '@SOURCE_DIR@'],
    output: 'doc',
    install: true,
    install_dir: get_option('datadir') / 'doc' / meson.project_name(),
  )
```

But unfortunately, LDoc gets confused and can't find the files it's looking for, like README.md, as it's looking for a relative path as specified in `config.ld`:

```lua
-- ...
topics = {"README.md"}
file = "lege"
examples = "examples"
```

I suppose I could canonicalize the paths in `config.ld`, but since Lua doesn't ship with a function for doing so, that may be more complex than it seems. In fact, with the above command LDoc can't even find `config.ld` -- I tried passing other flags to ldoc, such as `-b` and `-c`, which changed the behaviour somewhat (I was able to specify the exact path to `config.ld`), but in the end, LDoc just won't work unless it's run from the source directory.
